### PR TITLE
Add a proper sqlite3 database to RootPythia

### PR DESCRIPTION
**Pull request by ctmbl**
I replaced the `DummyDBManager` by a "real" `DatabaseManager` which initializes (in `_init_db`) and uses a database file `$DB_FOLDER/RootPythia.db`, configurable by environment variable. I also moved it to a separated `src/database` sub directory.

All SQL queries templates are stored in `database/db_structure.py` (which could as well be named `db_queries`), and the bot is safe from SQLi because I only used the sqlite3 safe substitution using `?` in queries.
The API hasn't change, apart from the class name, so the tests with slight changes still pass which gives me strong confidence in the bot working well.

The Database is currently only composed of one `users` table which stores exactly the same information that the previous python list stored. Improving this, is outside of the scope of this PR and will be done later as well as other bug fixes (rank and score that don't change when fetching new challenges). The `DatabaseManager` still returns `User`s objects to comply with other classes APIs, the conversion from the DB data to the `User` object (and vice versa) has been done thanks to slight additions to the `User` class to be developer-friendly.

I mainly used this tutorial: https://www.sqlitetutorial.net/sqlite-python/

### CODE REVIEW

**Comment by ctmbl** [comment]
@ZynoXelek could you take a look?

File: src/bot/root_pythia_bot.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 from api.rate_limiter import RateLimiter
 from bot.custom_help_command import RootPythiaHelpCommand
 from bot.root_pythia_cogs import RootPythiaCommands
-from bot.dummy_db_manager import DummyDBManager
+from database import DatabaseManager
```
Comment (ZynoXelek) [inline src/bot/root_pythia_bot.py:14]: Where does this import from ?

File: src/database/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from database.db_manager import DatabaseManager
```
Comment (ZynoXelek) [inline src/database/__init__.py:1]: Because here we specified the whole relative position

File: src/database/db_manager.py [hunk 3]
```diff
@@ -18,41 +29,74 @@ def __init__(self, idx=None, message=None):
...
+        self.logger = logging.getLogger(__name__)
+
+        self.DB_FOLDER = getenv("DB_FOLDER")
+        if self.DB_FOLDER is None or not path.isdir(self.DB_FOLDER):
+            self.logger.critical("DB_FOLDER: '%s', is not a directory", self.DB_FOLDER)
+            raise Exception(f"DB_FOLDER: '{self.DB_FOLDER}', is not a directory")
+
+        # Init Connection object allowing interaction with the database
+        db_file_path = path.join(self.DB_FOLDER, DB_FILE_NAME)
+        self.db = sqlite3.connect(db_file_path)
```
Comment (ZynoXelek) [inline src/database/db_manager.py:43]: Don't we need a `try except` here to detect the failed connection and write it in the logs ?

Comment (ZynoXelek) [inline src/database/db_manager.py:43]: Do we need a try except here to detect the failed connection and write it in the debugger ?

File: tests/conftest.py [hunk 4]
```diff
@@ -57,17 +57,19 @@ def mock_rootme_api_manager(mocker):
...
 @pytest.fixture
-def mock_dummy_db_manager(mock_rootme_api_manager):
+def mock_database_manager(mock_rootme_api_manager, monkeypatch, tmp_path):
     rootme_api_manager = mock_rootme_api_manager
+    monkeypatch.setenv("DB_FOLDER", str(tmp_path))
 
-    db = DummyDBManager(rootme_api_manager)
+    db = DatabaseManager(rootme_api_manager)
     yield db
+    db.db.close()
```
Comment (ZynoXelek) [inline tests/conftest.py:66]: Could you explain to me what those lines do ?
I didn't know this kind of code for tests before.

**Review by ZynoXelek — COMMENTED** [review]
What I didn't comment was clear and looks good to me.
Once my questions will be answered, I will approve this PR.

File: src/bot/root_pythia_bot.py (hunk 1 above)
Comment (ZynoXelek) [reply src/bot/root_pythia_bot.py:14]: (Compared to the import on the next comment I made)

File: tests/conftest.py (hunk 4 above)
Comment (ZynoXelek) [reply tests/conftest.py:66]: (I wanted to select the whole function here)

File: src/database/__init__.py (hunk 2 above)
Comment (ZynoXelek) [reply src/database/__init__.py:1]: Ok so we talked about it and in fact, the __init__.py is called by itself when importing database. That's why in the previous example, the DatabaseManager class can be imported.

File: src/bot/root_pythia_bot.py (hunk 1 above)
Comment (ZynoXelek) [reply src/bot/root_pythia_bot.py:14]: The solution is explained in the comment on the file `__init__.py`

File: src/database/db_manager.py (hunk 3 above)
Comment (ZynoXelek) [reply src/database/db_manager.py:43]: Well in fact, this is similar to a file opening so it should not fail. Moreover, if it does, it should rather crash to easily know something went wrong.

File: tests/conftest.py (hunk 4 above)
Comment (ZynoXelek) [reply tests/conftest.py:66]: The pytest function uses `fixtures` to create temporary objects to test on before deleting them.
So, in this `fixture`, we create a temporary dummy database. Then, we yield it for the tests to be done in the `test_*` functions.
Finally, this function is called once again at the end to process the remaining lines of the function (after the `yield`) and to delete the temporary created database.

**Review by ZynoXelek — APPROVED** [review]

**Pull request merged**